src/server/ua_server.c

```diff
@@ -693,6 +693,242 @@ UA_Server_removeCallback(UA_Server *server, UA_UInt64 callbackId) {
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_LOG_ERROR(server->config.logging, UA_LOGCATEGORY_SERVER,
+            "Add Certificates to CertificateGroup failed with error: %s."
+            "Either a CA certificate has no CRL, or a CRL was specified that"
+            "cannot be associated with a certificate.", UA_StatusCode_name(retval));
+        return retval;
+    }
+
+    /* When adding certificate files to the TrustList,
+     * it is not necessary to check the trust status of the existing SecureChannels. */
+    if(appendCertificates)
+        return certGroup->addToTrustList(certGroup, &trustList);
+
+    retval = certGroup->setTrustList(certGroup, &trustList);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    UA_DelayedCallback *dc = (UA_DelayedCallback*)UA_calloc(1, sizeof(UA_DelayedCallback));
+    if(!dc)
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+
+    dc->callback = secureChannel_delayedCloseTrustList;
+    dc->application = server;
+    dc->context = dc;
+
+    UA_EventLoop *el = server->config.eventLoop;
+    el->addDelayedCallback(el, dc);
+
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_StatusCode
+UA_Server_removeCertificates(UA_Server *server,
+                             const UA_NodeId certificateGroupId,
+                             UA_ByteString *certificates,
+                             size_t certificatesSize,
+                             const UA_Boolean isTrustedCertificates) {
+
+    UA_CertificateGroup *certGroup = getCertificateGroup(server, certificateGroupId);
+    if(!certGroup)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    UA_TrustListDataType oldTrustList;
+    UA_TrustListDataType_init(&oldTrustList);
+    UA_ByteString *crlsOld;
+    size_t crlsOldSize = 0;
+
+    if(isTrustedCertificates) {
+        oldTrustList.specifiedLists = UA_TRUSTLISTMASKS_TRUSTEDCRLS;
+        certGroup->getTrustList(certGroup, &oldTrustList);
+        crlsOld = oldTrustList.trustedCrls;
+        crlsOldSize = oldTrustList.trustedCrlsSize;
+    }
+    else {
+        oldTrustList.specifiedLists = UA_TRUSTLISTMASKS_ISSUERCRLS;
+        certGroup->getTrustList(certGroup, &oldTrustList);
+        crlsOld = oldTrustList.issuerCrls;
+        crlsOldSize = oldTrustList.issuerCrlsSize;
+    }
+
+    UA_ByteString *crls = NULL;
+    size_t crlsSize = 0;
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    size_t foundMatchCount = 0;
+    /* For each certificate to be removed, the corresponding CRL must also be deleted.
+     * If the certificate is not a CA certificate, the search can be skipped.
+     * However, if it is a CA certificate, all CRLs issued by that certificate must be removed.
+     * A CA certificate can also have multiple CRLs. */
+    for(size_t i = 0; i < certificatesSize; i++) {
+        UA_Boolean foundMatch = false;
+        for(size_t j = 0; j < crlsOldSize; j++) {
+            retval = UA_CertificateUtils_ckeckCrlMatch(&certificates[i], &crlsOld[j]);
```

Comment:
I understand the intention to process everything as high up in the stack as possible to do it only at one location.
But this specific check here now involves parsing the same CRLs multiple times. There are already also other cases regarding the certificate validation where a piece of work internally requires doing conversions multiple times. It feels like offloading those tasks to the specific crypto layers would not be a bad solution to at least reduce the duplication of conversion work.

I may also add that in an old discussion we also had the topic of what the removeFromTrustList callback should be responsible for. And back then it had the responsibility to also remove all CRLs as required (which basically meant that the checking part should be offloaded into the crypto layer).